File: extra-problems/1dDpProblems.js
/*
old problems:
-climb stairs
-chicken nuggets
-coin change
-jump game

-longest increasing subsequence
-valid perfect squares

-word break
-work break ii
*/

/*
longest alternating subsequence (+, -, +, -, ...)

input = [10, 13, -8, 4, 8, -9, 224]
length of longest = 5
one example: [13, -8 ,4, -9, 224]

don't use this
*/

/*
longest doubling sequence (n, 2n, 4n, 8n, ..., (2^i)*n)

input = [93,4,3,17,186,34,290,813,372,8,6]
length of longest = 3
example longest sequence: [93, 186, 372]

{
  93: 1
  4: 1
  3: 1
  17: 1
  186: 2
  34: 1
}
*/

const LENGTH_OF_SEQUENCE_WITH_ONE_NUMBER = 1;

const isEven = (num) => num % 2 === 0;

const getObjectValueWithDefault = (object, key, defaultValue) => {
  const keyInObject = object.hasOwnProperty(key);

  if (keyInObject) return object[key];

  return defaultValue;
};

const getMaxObjectValue = (object) => {
  const values = Object.values(object);

  return Math.max(...values);
};

const findLongestDoublingSequence = (nums) => {
  const longestDoublingSequenceEndingAt = {};

  nums.forEach((num) => {
    if (!isEven(num)) {
      longestDoublingSequenceEndingAt[num] = 1;
      return;
    }

    const prevNumInSequence = num / 2;
    const longestPriorSequenceLength = getObjectValueWithDefault(
      longestDoublingSequenceEndingAt,
      prevNumInSequence,
      LENGTH_OF_SEQUENCE_WITH_ONE_NUMBER,
    );

    longestDoublingSequenceEndingAt[num] = longestPriorSequenceLength + 1;
  });

  return getMaxObjectValue(longestDoublingSequenceEndingAt);
};

// const input = [93, 4, 3, 17, 186, 34, 290, 813, 372, 8, 6];
// console.log(findLongestDoublingSequence(input));

/*
at top of building with height h
need to reach floor at height 0

at each height, there are fixed set of options for how far you can jump down
(e.g. if we have [1,3], then you CANNOT jump down 2 floors)

return min jumps to reach ground floor


[[],   [1], [5], [1,2] [1,3], [1,2]]
 0      1    2    3     4      5

h=5

to be done, index must be <= 0

example output: 3 (5 => 3 => 2 => 0)
*/

// const GROUND_FLOOR = 0;
// const CANNOT_REACH_GROUND_FLOOR = -1;

// const createArrayOfSize = (size, fillValue) => {
//   return new Array(size).fill(fillValue);
// };

// const checkIfCanReachCurFromAbove = (jumpOptions, curFloor, aboveFloor) => {
//   const aboveJumpOptions = jumpOptions[aboveFloor];

//   for (const jumpSize of aboveJumpOptions) {
//     if (aboveFloor - jumpSize === curFloor) return true;
//   }

//   return false;
// };

// const getMinJumpsToGroundFloor = (jumpOptions) => {
//   const numFloors = jumpOptions.length;
//   const minJumpsToReach = createArrayOfSize(numFloors, Infinity);

//   const lastFloor = numFloors - 1;
//   minJumpsToReach[lastFloor] = 0;

//   for (let curFloor = lastFloor - 1; curFloor >= 0; curFloor--) {
//     for (let aboveFloor = curFloor + 1; aboveFloor < numFloors; aboveFloor++) {
//       const canReachCurFromAbove = checkIfCanReachCurFromAbove(
//         jumpOptions,
//         curFloor,
//         aboveFloor,
//       );

//       if (canReachCurFromAbove) {
//         minJumpsToReach[curFloor] = Math.min(
//           minJumpsToReach[curFloor],
//           1 + minJumpsToReach[aboveFloor],
//         );
//       }
//     }
//   }

//   const canReachGroundFloor = minJumpsToReach[GROUND_FLOOR] !== Infinity;
//   if (!canReachGroundFloor) return CANNOT_REACH_GROUND_FLOOR;

//   return minJumpsToReach[GROUND_FLOOR];
// };

// const input = [[], [1], [5], [1, 2], [1, 3], [1, 2]];
// console.log(getMinJumpsToGroundFloor(input));

/*
multiplicative jump game

array will be one-indexed

from index i, you can jump to index i * nums[i] (not any earlier though)
can start at index 1

return min jumps to reach last index or -1 if not possible

example:
input = [3,2,3,1,7,2,4,1,8]
         1 2 3 4 5 6 7 8 9
output = 2 (1 => 3 => 9)
*/

// const START_LOCATION = 1;
// const CANNOT_REACH_END = -1;

// const getZeroIndexedValue = (index) => index - 1;

// const createArrayOfSize = (size, fillValue) => {
//   return new Array(size).fill(fillValue);
// };

// const getMinJumpsToEnd = (nums) => {
//   const minJumpsToReach = createArrayOfSize(nums.length, Infinity);
//   minJumpsToReach[getZeroIndexedValue(START_LOCATION)] = 0;

//   for (let curIdx = START_LOCATION + 1; curIdx <= nums.length; curIdx++) {
//     for (let prevIdx = 1; prevIdx < curIdx; prevIdx++) {
//       const reachableIdx = prevIdx * nums[getZeroIndexedValue(prevIdx)];

//       if (reachableIdx === curIdx) {
//         const newJumpsToReachCur =
//           minJumpsToReach[getZeroIndexedValue(prevIdx)] + 1;
//         const prevMinJumps = minJumpsToReach[getZeroIndexedValue(curIdx)];

//         minJumpsToReach[getZeroIndexedValue(curIdx)] = Math.min(
//           prevMinJumps,
//           newJumpsToReachCur,
//         );
//       }
//     }
//   }

//   const lastIdx = nums.length - 1;
//   const canReachEnd = minJumpsToReach[lastIdx] !== Infinity;

//   if (!canReachEnd) return CANNOT_REACH_END;

//   return minJumpsToReach[lastIdx];
// };

// const input = [3, 2, 3, 1, 7, 2, 4, 1, 8];
// console.log(getMinJumpsToEnd(input));

File: extra-problems/1d_dp_problems.py
"""
PROBLEM 1:
longest doubling sequence (n, 2n, 4n, 8n, ..., (2^i)*n)
there are no duplicate nums

input = [93,4,3,17,186,34,290,813,372,8,6]
length of longest = 3
example longest sequence: [93, 186, 372]

{
  93: 1
  4: 1
  3: 1
  17: 1
  186: 2
  34: 1
}
"""


def find_longest_doubling_seq_length(nums):
    n = len(nums)
    longest_ending_with = [1 for _ in range(n + 1)]
    
    for i in range(n):
        for prev_idx in range(i):
            if nums[prev_idx] * 2 == nums[i]:
                longest_ending_with[i] = max(longest_ending_with[i], 1 + longest_ending_with[prev_idx])

    return max(longest_ending_with)

# print(find_longest_doubling_seq_length([93,4,3,17,186,34,290,813,372,8,6]))

"""
PROBLEM 2:
at top of building with height h
need to reach floor at height 0

at each height, there are fixed set of options for how far you can jump down
(e.g. if we have [1,3], then you CANNOT jump down 2 floors)

return min jumps to reach ground floor (if not possible, return -1)


[[],   [1], [5], [1,2] [1,3], [1,2]]
 0      1    2    3     4      5

h=5

to be done, index must be <= 0

example output: 3 (5 => 3 => 2 => 0)
"""

# def min_jumps_to_ground(height, jump_sizes, cache={}):
#     assert len(jump_sizes) >= height + 1

#     if height <= 0:
#         return 0
    
#     if height in cache:
#         return cache[height]

#     min_jumps = float('inf')

#     for jump_size in jump_sizes[height]:
#         min_jumps = min(min_jumps, 1 + min_jumps_to_ground(height - jump_size, jump_sizes, cache))

#     # Doesn't properly handle the case where it's possible. Would
#     # need to modify this
#     cache[height] = min_jumps
#     return min_jumps

def min_jumps_to_ground(height, jump_sizes):
    min_jumps_from = [float('inf') for _ in range(height + 1)]
    min_jumps_from[0] = 0

    for h in range(1, height + 1):
        for jump_size in jump_sizes[h]:
            if h - jump_size <= 0:
                min_jumps_from[h] = 1
                break

            min_jumps_from[h] = min(min_jumps_from[h], 1 + min_jumps_from[h - jump_size])

    return min_jumps_from[height]

print(min_jumps_to_ground(5, [[],   [1], [5], [1,2], [1,3], [1,2]]))


